docs/connections.md

```diff
@@ -0,0 +1,597 @@
+```
+
+If not for this second rule, the `a` would be visible only in place of
+expression `b`. However, now it is visible in the outer lambda body and can be
+accessed. The only externally provided identifier must be `default` method.
+
+--- 
+Lambdas may not appear in the assignment's pattern (i.e. values cannot be
+pattern-matched into lambda).
+
+So the following is not valid:
+```
+a -> b = foo
+```
+
+---
+Example:
+```
+a -> a -> b
+```
+
+Here the first `a` and second `a` are separate identifiers, the latter shadowing
+the first one. If one wanted to express that both arguments are of the same
+type, `a -> A -> b` would have been used. `b` refers to an identifier from
+graph's scope (it is in the body's position, not pattern).
+
+OPEN QUESTION: actually it might be "nice" to have both `a` unified in such
+case. 
+
+
+
+## Type ascription
+The type ascription operator `:` introduces pattern context for its right hand
+side. The basic form is `value:type`. It says that `value` be of the given
+`type`, i.e. that all its possible values belong to the set of atoms represented
+by type.
+
+The effect of this can be two-fold. If `value` is of (at least partially) known
+type, appropriate constraints will be introduced on the types denoted by
+variable names appearing in the pattern context. If the identifier was not
+defined, it will appear in the current scope.
+
+For example:
+```
+5 : «a»?
+```
+
+This introduces constraint on the type `a` that its value set must include atom
+`5`. If the `a` is already visible in the scope, this constraint will be added.
+Otherwise, `a` will be introduced into the current scope with that constraint.
+
+// TODO: What if parent scope only ascribes identifier with type constraint but
+only nested scope assigns to it? 
+
+// TODO: Open design question, if `a` should modify existing variable or should
+always try to shadow it. What is the difference between `5 : a` and `5 : A`?
+(except the latter not being able to ever introduce a new identifier)
+
+// TODO: Is it legal or sensible to ascribe a variable when there is no binding?
+
+
+When type is known, the type ascription can be used to constrain type of the
+value:
+
+```
+»a« : 5
+```
+
+This says that value of `a` is of type `5`. Type `5` has only a single allowed
+value: `5`. This will tell compiler to error out if program tries to bind `5`
+with any value that is not known to be `5`.
+
+However, this example refers to some `a` already being visible in scope and does
+not introduce any identifiers.
+
+The `type` in this expression is pattern context and can be used to constrain
+the type variables. It is legal to assign constraints on an identifier using `:`
+multiple times in any of the scopes where identifier is visible.
+
+Signatures are just type ascriptions that happen to precede the assignments.
+They have no special rules currently defined. This area needs further design
+work.
+
+TODO: Open question: does empty type (`Void`) exists in the language?
```

Comment:
We do have a `Unit` type, yes.

Reply:
Ah, I see what you mean. I don't think I've addressed that question in the type system doc yet. When we merge docs I'll make sure to add it as a TODO item.